### PR TITLE
Add `GroupedQueryAttention` layer

### PR DESCRIPTION
**Pull request by awsaf49**
This PR corresponds to Issue: https://github.com/keras-team/keras/issues/18402.

### CODE REVIEW

**Comment by awsaf49** [comment]
Hello @mattdangerw,

I would appreciate your valuable feedback on the implementation. I used [llamav2](https://github.com/facebookresearch/llama/blob/main/llama/model.py) as a reference for this layer but also used `keras.layers.MultiHeadAttention` to maintain the `keras-api`.

I did observe some key differences between the llamav2 implementation and the Keras `MultiHeadAttention`:

1. Llamav2 approaches this layer as **self-attention**, accepting a single tensor as input, whereas Keras `MultiHeadAttention` takes query, key, and value as separate inputs.

2. In Llamav2, the layer accepts a `mask` input and directly integrates it with `attention_scores`. In contrast, Keras MHA computes the mask internally within the layer and applies it separately within the `Softmax` layer.

3. Llamav2 utilizes a **key-value** `cache` mechanism, which differs somewhat from `keras-nlp.layers.CachedMultiHeadAttention`. I'm currently undecided whether to incorporate the cache within this layer or create a separate layer akin to the `keras-nlp` approach. Your insights on this matter would be highly appreciated.

Additionally, I haven't yet included the `Dense` layer `kwargs` and unit testing. I believe this can be easily integrated at a later stage.

File: keras/layers/attention/group_query_attention.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+    Returns:
+        attention_output: Result of the computation, of shape `(B, T, dim)`,
+            where `T` is for target sequence shapes and `dim` is the query
+            input last dim.
+        attention_scores: (Optional) attention coefficients.
+    """
+
+    def __init__(
+        self, head_dim, num_q_heads, num_kv_heads, use_bias=False, **kwargs
```
Comment (fchollet) [inline keras/layers/attention/group_query_attention.py:60]: We should not use abbreviations in layer signatures, e.g. `num_query_heads` rather than `num_q_heads`

File: keras/layers/attention/group_query_attention.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        query,
+        value,
+        key=None,
+        mask=None,
+        return_attention_scores=False,
+        training=None,
+    ):
+        if key is None:
+            key = value
+        B, T, _ = ops.shape(query)  # (B, T, dim)
```
Comment (fchollet) [inline keras/layers/attention/group_query_attention.py:118]: Use a fully-spelled out snake case name like `batch_dim`

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/layers/attention/group_query_attention.py (hunk 1 above)
Comment (awsaf49) [reply keras/layers/attention/group_query_attention.py]: I have updated it

File: keras/layers/attention/group_query_attention.py (hunk 2 above)
Comment (awsaf49) [reply keras/layers/attention/group_query_attention.py]: Ok. Btw, I have also used fully-spelled names in comments and docstring, I hope that's okay.

File: keras/layers/attention/group_query_attention.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
+from keras import ops
```
Comment (mattdangerw) [inline keras/layers/attention/group_query_attention.py:1]: Make sure to add tests!

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some initial high level comments.

File: keras/layers/attention/group_query_attention.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+    def build(
+        self,
+        query_shape,
+        value_shape,
+        key_shape=None,
+    ):
+        key_shape = value_shape if key_shape is None else key_shape
+        self.feature_dim = query_shape[-1]
+        self._query_dense = Dense(
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras/layers/attention/group_query_attention.py:87]: If it's useful. Here's a draft on a multi-query attention build block I was using for a model port. This would need adapting for grouped query (key and value should still output 4d).

```python
        # Einsum variables:
        # b = batch size
        # q = query length
        # k = key/value length
        # m = model dim
        # n = num heads
        # h = head dim
        model_dim = query_shape[-1]
        head_dim = model_dim // self.num_heads
        key_shape = value_shape if key_shape is None else key_shape
        self.query_dense = keras.layers.EinsumDense(
            "bqm,mnh->bqnh",
            output_shape=(None, self.num_heads, head_dim),
            kernel_initializer="variance_scaling",
            name="query_dense",
        )
        self.query_dense.build(query_shape)
        self.value_dense = keras.layers.EinsumDense(
            "bkm,mh->bkh",
            output_shape=(None, head_dim),
            kernel_initializer="variance_scaling",
            name="value_dense",
        )
        self.value_dense.build(value_shape)
        self.key_dense = keras.layers.EinsumDense(
            "bkm,mh->bkh",
            output_shape=(None, head_dim),
            kernel_initializer="variance_scaling",
            name="key_dense",
        )
        self.key_dense.build(key_shape)
        self._dot_product_equation = "bqnh,bkh->bnqk"
        self._combine_equation = "bnqk,bkh->bqnh"
        self.output_dense = keras.layers.EinsumDense(
            "bqnh,nhm->bqm",
            output_shape=(None, model_dim),
            name="output_dense",
        )
        self.output_dense.build((None, None, self.num_heads, head_dim))
        self.softmax_layer = keras.layers.Softmax()
        self.built = True
```

[the remark this answers is not in the document]
Comment (awsaf49) [reply keras/layers/attention/group_query_attention.py:87]: Thank you. I apologize for the delayed response. Unfortunately, I was caught up in a conference deadline.

[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras/layers/attention/group_query_attention.py:87]: All good! No rush. Just realized I had some code laying around that might be useful here.

**Comment by fchollet** [comment]
The code / API looks good to me (would like @mattdangerw to LGTM it though). Please add unit tests!

File: keras/layers/attention/group_query_attention.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (awsaf49) [reply keras/layers/attention/group_query_attention.py]: @mattdangerw Apologies for the delay once again; I've just returned home after attending the conference abroad.

I have updated the code to incorporate all your feedback. Thanks for the code snippet; they were helpful.

Since this is my first time working with `einsum`, it would be helpful if you could double-check the equations and also confirm if the `bias_axes` are correct.

Next, I will add the unit tests. Additionally, I couldn't find the instructions for `bazel` tests that used to be in this repository. Could you please point me to that page?

[the remark this answers is not in the document]
Comment (awsaf49) [reply keras/layers/attention/group_query_attention.py]: @mattdangerw  unit tests have also been added.

**Review by mattdangerw — APPROVED** [review]
Sorry for the delay! Overall this looks good to me, just a few last comments.

Do we know of a reference layer somewhere in a torch or jax lib we could use to sanity check our numerics?

File: keras/layers/attention/grouped_query_attention_test.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+import pytest
+from absl.testing import parameterized
+
+from keras import backend
+from keras import initializers
+from keras import layers
+from keras import testing
+
+
+class GroupedQueryAttentionTest(testing.TestCase, parameterized.TestCase):
```
[the remark this answers is not in the document]
Comment (awsaf49) [reply keras/layers/attention/grouped_query_attention_test.py:11]: There is already test with MQA (`num_key_value_head=1`) and GQA (`num_key_value_head < num_query_head`) [here](https://github.com/awsaf49/keras/blob/a3b89dc2a00c910bfad4c9b79b11c892b9c46c06/keras/layers/attention/grouped_query_attention_test.py#L54-L55), should I also add another test on `basics` ?

File: keras/layers/attention/grouped_query_attention.py [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+        query_shape,
+        value_shape,
+        key_shape=None,
+    ):
+        # Einsum variables:
+        # b = batch size
+        # q = query length
+        # k = key/value length
+        # m = model dim
+        # n = num heads
```
[the remark this answers is not in the document]
Comment (awsaf49) [reply keras/layers/attention/grouped_query_attention.py:123]: Yes I was having the same thoughts but couldn't come up with sth at that time. I'll try again

File: keras/layers/attention/grouped_query_attention_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras/layers/attention/grouped_query_attention_test.py:11]: Ah I see, I thought that test was just `compute_output_shape`, but it also exercises call. I think that's all good!

File: keras/layers/attention/grouped_query_attention.py [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+        kernel_constraint=None,
+        bias_constraint=None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.supports_masking = True
+        self.head_dim = head_dim
+        self.num_query_heads = num_query_heads
+        self.num_key_value_heads = num_key_value_heads
+        self.num_repeats = num_query_heads // num_key_value_heads
```
[the remark this answers is not in the document]
Comment (awsaf49) [reply keras/layers/attention/grouped_query_attention.py:107]: added the divisible check

File: keras/layers/attention/grouped_query_attention.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (awsaf49) [reply keras/layers/attention/grouped_query_attention.py]: I've used `u` for query heads and used `v` for key-value heads.

File: keras/layers/attention/grouped_query_attention.py [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+        use_causal_mask: A boolean to indicate whether to apply a causal mask to
+            prevent tokens from attending to future tokens (e.g., used in a
+            decoder Transformer).
+
+    Returns:
+        attention_output: Result of the computation, of shape
+            `(batch_dim, target_seq_len, feature_dim)`, where `target_seq_len`
+            is for target sequence length and `feature_dim` is the query input
+            last dim.
+        attention_scores: (Optional) attention coefficients.
```
[the remark this answers is not in the document]
Comment (awsaf49) [reply keras/layers/attention/grouped_query_attention.py]: done

**Comment by awsaf49** [comment]
> Sorry for the delay! Overall this looks good to me, just a few last comments.
> 
> Do we know of a reference layer somewhere in a torch or jax lib we could use to sanity check our numerics?

I tried to match numeric result of pytorch (simple ver. of llama-v2) to this impl. but I am having trouble initialing them with same weights which I think is essential for testing. Probably it's due to `Einsum`.

**Comment by fchollet** [comment]
> I tried to match numeric result of PyTorch (simple ver. of llama-v2) with this impl. but I am having trouble initialing them with same weights which I think is essential for testing. Probably it's due to Einsum.

Do you mean the weight layout is different? What are the different weight matrixes and their shapes?

**Comment by awsaf49** [comment]
> > I tried to match numeric result of PyTorch (simple ver. of llama-v2) with this impl. but I am having trouble initialing them with same weights which I think is essential for testing. Probably it's due to Einsum.
> 
> Do you mean the weight layout is different? What are the different weight matrixes and their shapes?

@fchollet  Thanks or your help with this query! After some digging, I noticed an interesting difference in the weight shapes between `Linear` layer in PyTorch version and the `EinsumDense` layer in Keras. In PyTorch, the weight shape for the `Linear` layer looks like `(n_q_heads * head_dim, dim)`, while for the `EinsumDense` layer in Keras, it's `(dim, n_q_heads, head_dim)`. To make things work, I had to add a little `reshape` action after the `transpose` operation to fix the shape mismatch.

After that, I ran a numeric comparison of GQA between Keras (TensorFlow backend) and PyTorch version (code above) and it matched up perfectly with the PyTorch. I used the following code to test,

```py
# Define parameters and data
head_dim = 2
n_q_heads = 3 * 2
n_kv_heads = 3
batch_size = 2
seq_len = 4
dim = 16

# Instantiate the torch GQA
gqa_torch = GroupedQueryAttentionTorch(dim=dim, 
                                       head_dim=head_dim,
                                       n_kv_heads=n_kv_heads, 
                                       n_q_heads=n_q_heads)


# Manually initialize PyTorch model weights to known values
init_tensor1 = torch.rand(n_q_heads * head_dim, dim)
init_tensor2 = torch.rand(n_kv_heads * head_dim, dim)
init_tensor3 = torch.rand(dim, n_q_heads * head_dim)
gqa_torch.wq.weight = nn.Parameter(init_tensor1.clone())
gqa_torch.wk.weight = nn.Parameter(init_tensor2.clone())
gqa_torch.wv.weight = nn.Parameter(init_tensor2.clone())
gqa_torch.wo.weight = nn.Parameter(init_tensor3.clone())

# Generate some example input data
inputs = torch.randn(batch_size, seq_len, dim)
mask = None  # You can add a mask tensor here if needed

# Forward Pass
output_torch = gqa_torch(inputs, mask).detach().numpy()
print(output_torch.shape)


# Instantiate the keras GQA
gqa_keras = GroupedQueryAttentionKeras(head_dim=head_dim,
                                    num_query_heads=n_q_heads,
                                    num_key_value_heads=n_kv_heads)
gqa_keras.build((batch_size, seq_len, dim), (batch_size, seq_len, dim))

# Manually initialize Keras model weights to same known values
gqa_keras._query_dense.kernel = tf.Variable(init_tensor1.numpy().T.reshape(dim, n_q_heads, head_dim))
gqa_keras._key_dense.kernel = tf.Variable(init_tensor2.numpy().T.reshape(dim, n_kv_heads, head_dim))
gqa_keras._value_dense.kernel = tf.Variable(init_tensor2.numpy().T.reshape(dim, n_kv_heads, head_dim))
gqa_keras._output_dense.kernel = tf.Variable(init_tensor3.numpy().T.reshape(n_q_heads, head_dim, dim))


# Generate inputs
inputs = tf.convert_to_tensor(inputs.numpy())
mask = None  # You can add a mask tensor here if needed

# Forward Pass
output_keras = gqa_keras(inputs, inputs, attention_mask=mask).numpy()
print(output_keras.shape)

# Test if torch keras result match
np.testing.assert_allclose(output_keras, output_torch, rtol=1e-5, atol=1e-5)
```

**Review by fchollet — APPROVED** [review]
LGTM -- thank you for the great contribution! 👍

**Comment by awsaf49** [comment]
Thanks for your feedbacks @mattdangerw and @fchollet
